### PR TITLE
add nori_number token filter in analysis-nori

### PR DESCRIPTION
**Pull request by danmuzi**
The `KoreanNumberFilter` has included in Nori after Lucene 8.2.0. ([LUCENE-8812](https://issues.apache.org/jira/browse/LUCENE-8812))
However, it isn't supported now in Nori Analysis plugin. (Kuromoji supports `kuromoji_number`)
It seems to be missing(https://github.com/elastic/elasticsearch/pull/30397) because `KoreanNumberFilter` didn't exist at Lucene 7.4.0 that supports Nori first time.
This PR is for that.

### CODE REVIEW

File: docs/plugins/analysis-nori.asciidoc [hunk 1]
```diff
@@ -434,3 +434,52 @@ Which responds with:
 --------------------------------------------------
 
 <1> The Hanja form is replaced by the Hangul translation.
+
+
+[[analysis-nori-number]]
+==== `nori_number` token filter
```
Comment (jimczi) [inline docs/plugins/analysis-nori.asciidoc]: We also need to expose `discard_punctuation` in the tokenizer in order to detect floating point numbers and to split numbers correctly if they are separated with whitespaces ? We should then advise in the documentation to set `discard_punctuation` to `false` when using the `nori_number` filter and to remove punctuations in a subsequent filter ?

**Review by jimczi — COMMENTED** [review]
Thanks for adding this filter `danmuzi` ! I left one comment regarding the `discard_punctuation` option of the tokenizer that was added to handle the number filter correctly.

File: docs/plugins/analysis-nori.asciidoc (hunk 1 above)
Comment (danmuzi) [reply docs/plugins/analysis-nori.asciidoc]: I think so. 
I added more description for that.
Please check the additional commits.

**Comment by danmuzi** [comment]
Thanks for your review, @jimczi  👍 
I added some commits reflecting your comments.
But I'm not sure it's right to include the `discard_punctuation` option in this PR.
Because this PR is for `nori_number` token filter.
So I submitted commits separately.
If you think creating a new PR for `discard_punctuation` is better, I will write a new PR for that.
If so, I'll remove the `discard_punctuation` commit in this PR and rebase to master branch after `discard_punctuation` PR is submitted.
What do you think about this?

File: plugins/analysis-nori/src/test/java/org/elasticsearch/index/analysis/NoriAnalysisTests.java [hunk 2]
```diff
@@ -159,6 +162,20 @@ public void testNoriReadingForm() throws IOException {
...
+            .put("index.analysis.filter.my_filter.type", "nori_number")
+            .build();
+        TestAnalysis analysis = AnalysisTestsHelper.createTestAnalysisFromSettings(settings, new AnalysisNoriPlugin());
+        TokenFilterFactory factory = analysis.tokenFilter.get("my_filter");
+        Tokenizer tokenizer = new KoreanTokenizer();
+        tokenizer.setReader(new StringReader("오늘 십만이천오백원짜리 와인 구입"));
+        TokenStream stream = factory.create(tokenizer);
+        assertTokenStreamContents(stream, new String[] {"오늘", "102500", "원", "짜리", "와인", "구입"});
+    }
+
```
Comment (jimczi) [inline plugins/analysis-nori/src/test/java/org/elasticsearch/index/analysis/NoriAnalysisTests.java]: Can you add a small test for the new discard_punctuation option of the tokenizer ?

File: docs/plugins/analysis-nori.asciidoc [hunk 3]
```diff
@@ -434,3 +439,105 @@ Which responds with:
...
+
+The `nori_number` token filter normalizes Korean numbers
+to regular Arabic decimal numbers in half-width characters.
+
+Korean numbers are often written using a combination of Hangul and Arabic numbers with various kinds punctuation.
+For example, ３．２천 means 3200.
+This filter does this kind of normalization and allows a search for 3200 to match ３．２천 in text,
+but can also be used to make range facets based on the normalized numbers and so on.
+
+Notice that this analyzer uses a token composition scheme and relies on punctuation tokens
```
Comment (jimczi) [inline docs/plugins/analysis-nori.asciidoc]: Maybe add add a `NOTE:` to emphasize this part ?

**Review by jimczi — COMMENTED** [review]
I left two minor comments but the change looks good to me.

> But I'm not sure it's right to include the discard_punctuation option in this PR.
Because this PR is for nori_number token filter.

I think it's ok since the discard_punctuation option was added specifically for the number token filter. Let's add both in the same pr, thanks for separating the commits though.

**Comment by danmuzi** [comment]
Thanks Jim.
I added a commit based on your review.
It's about adding NOTE in asciidoc and test cases for discard_punctuation.
Please check it!

File: docs/plugins/analysis-nori.asciidoc (hunk 3 above)
Comment (danmuzi) [reply docs/plugins/analysis-nori.asciidoc]: Done!

File: plugins/analysis-nori/src/test/java/org/elasticsearch/index/analysis/NoriAnalysisTests.java (hunk 2 above)
Comment (danmuzi) [reply plugins/analysis-nori/src/test/java/org/elasticsearch/index/analysis/NoriAnalysisTests.java]: Done!

**Comment by jimczi** [comment]
@elasticmachine ok to test

**Comment by danmuzi** [comment]
I'm not sure why elasticsearch-ci/2 and elasticsearch-ci/bwc and elasticsearch-ci/default-distro are failed.
It doesn't seem related but I found an indentation problem in this PR.
And I rebased this patch to the latest master branch commit.

**Comment by danmuzi** [comment]
Because of the my rebase mistake, the previous Jenkins build history has been lost in this conversation.
So I attach the failure builds before the rebase.
elasticsearch-ci/2 : https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-2/19332/
elasticsearch-ci/bwc : https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-bwc/19021/
elasticsearch-ci/default-distro : https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request+default-distro/18876/

After the rebase, all Jenkins builds passed.

**Review by jimczi — APPROVED** [review]
LGTM, thanks @danmuzi !

**Comment by danmuzi** [comment]
Thanks for your kind reviews! @jimczi

**Pull request merged**